pkg/template/processor.go

```diff
@@ -0,0 +1,193 @@
+	return Processor{cl: cl, scheme: scheme}
+}
+
+// Process processes the template (ie, replaces the variables with their actual values) and optionally filters the result
+// to return a subset of the template objects
+func (p Processor) Process(tmpl *templatev1.Template, values map[string]string, filters ...FilterFunc) ([]runtime.RawExtension, error) {
+	// inject variables in the twmplate
+	for param, val := range values {
+		v := templateprocessing.GetParameterByName(tmpl, param)
+		if v != nil {
+			v.Value = val
+			v.Generate = ""
+		}
+	}
+	// convert the template into a set of objects
+	tmplProcessor := templateprocessing.NewProcessor(map[string]generator.Generator{
+		"expression": generator.NewExpressionValueGenerator(rand.New(rand.NewSource(time.Now().UnixNano()))),
+	})
+	if err := tmplProcessor.Process(tmpl); len(err) > 0 {
+		return nil, errs.Wrap(err.ToAggregate(), "unable to process template")
+	}
+	var result templatev1.Template
+	if err := p.scheme.Convert(tmpl, &result, nil); err != nil {
+		return nil, errs.Wrap(err, "failed to convert template to external template object")
+	}
+	return Filter(result.Objects, filters...), nil
+}
+
+// Apply applies the objects, ie, creates or updates them on the cluster
+func (p Processor) Apply(objs []runtime.RawExtension) error {
+	for _, rawObj := range objs {
+		obj := rawObj.Object
+		if obj == nil {
+			continue
+		}
+		gvk := obj.GetObjectKind().GroupVersionKind()
+		_, err := p.ApplySingle(obj, true, nil)
+		if err != nil {
+			return errs.Wrapf(err, "unable to create resource of kind: %s, version: %s", gvk.Kind, gvk.Version)
+		}
+	}
+	return nil
+}
+
+// ApplySingle creates the object if is missing and if the owner object is provided, then it's set as a controller reference.
+// If the objects exists then based on the UpdateStrategy it's updated or let as it is and updated is skipped.
+// The return boolean says if the object was either created or updated
+func (p Processor) ApplySingle(obj runtime.Object, forceUpdate bool, owner v1.Object) (bool, error) {
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	createdOrUpdated, err := p.createOrUpdateObj(obj, forceUpdate, owner)
+	if err != nil {
+		return createdOrUpdated, errs.Wrapf(err, "unable to create resource of kind: %s, version: %s", gvk.Kind, gvk.Version)
+	}
+	return createdOrUpdated, nil
+}
+
+func (p Processor) createOrUpdateObj(newResource runtime.Object, forceUpdate bool, owner v1.Object) (bool, error) {
+	// gets the meta accessor to the new resource
+	metaNew, err := meta.Accessor(newResource)
+	if err != nil {
+		return false, errors.Wrapf(err, "cannot get metadata from %+v", newResource)
+	}
+
+	// creates a deepcopy of the new resource to be used to check if it already exists
+	existing := newResource.DeepCopyObject()
+
+	// set current object as annotation
+	//newWithoutAnnotation := newResource.DeepCopyObject()
+	annotations := metaNew.GetAnnotations()
+	newConfiguration := getNewConfiguration(newResource)
```

Comment:
you don't really need to assign the result of `getNewConfiguration(newResource)` (that's a minor thing, though)

Reply:
Yes, I do because it's then used here https://github.com/codeready-toolchain/toolchain-common/pull/64/files#diff-ae1e62ddb6744c2036a1b470ddc6a527R132

Reply:
ah ok, I was missing the rest of the func body :/